File: apps/medousa-home/src/lib/stream/reconnect.ts
/**
 * Client-side reconnect discipline for Tauri SSE streams.
 * Mirrors medousa-sdk Rust/Python reconnect helpers: bounded backoff,
 * overlap guard, circuit breaker, and spine replay via `?since=<seq>`.
 */

export const MAX_STREAM_RECONNECT_DELAY_MS = 30_000;

export interface BackoffPolicy {
  baseMs: number;
  factor: number;
  maxMs: number;
  maxAttempts: number | null;
}

export interface CircuitBreakerConfig {
  /** Consecutive failures that trip the breaker Closed -> Open. */
  failureThreshold: number;
  /** How long the breaker stays Open before allowing a half-open probe. */
  cooldownMs?: number;
  /** Consecutive half-open successes required to fully close again. */
  successThreshold?: number;
}

/** Default cooldown before a tripped breaker allows a half-open probe. */
export const DEFAULT_BREAKER_COOLDOWN_MS = 15_000;

export type CircuitState = "closed" | "open" | "half_open";

export const DEFAULT_INTERACTIVE_BACKOFF: BackoffPolicy = {
  baseMs: 500,
  factor: 2,
  maxMs: MAX_STREAM_RECONNECT_DELAY_MS,
  maxAttempts: 10,
};

export const DEFAULT_WORKSPACE_BACKOFF: BackoffPolicy = {
  baseMs: 1_000,
  factor: 2,
  maxMs: MAX_STREAM_RECONNECT_DELAY_MS,
  maxAttempts: null,
};

export function reconnectDelayMs(policy: BackoffPolicy, attempt: number): number {
  const raw = Math.min(policy.baseMs * policy.factor ** attempt, policy.maxMs);
  const jitter = 0.5 + ((attempt * 7919) % 500) / 1000;
  return Math.round(raw * jitter);
}

export function mayRetry(policy: BackoffPolicy, attempt: number): boolean {
  if (policy.maxAttempts == null) return true;
  return attempt < policy.maxAttempts;
}

/** Append or replace `?since=` for durable spine replay. */
export function streamPathWithSince(path: string, since: number): string {
  const base = path.split("?")[0] ?? path;
  if (since <= 0) return base;
  return `${base}?since=${since}`;
}

export interface SeqTrackedEvent {
  turn_id: string;
  seq?: number;
}

/** Returns false when the event is a duplicate (seq <= last seen). */
export function applyStreamSeq(
  lastSeqByTurn: Map<string, number>,
  event: SeqTrackedEvent,
): boolean {
  const seq = event.seq ?? 0;
  if (seq <= 0) return true;
  const last = lastSeqByTurn.get(event.turn_id) ?? 0;
  if (seq <= last) return false;
  lastSeqByTurn.set(event.turn_id, seq);
  return true;
}

export class OverlapGuard {
  private active = false;

  tryEnter(): boolean {
    if (this.active) return false;
    this.active = true;
    return true;
  }

  release(): void {
    this.active = false;
  }

  get isActive(): boolean {
    return this.active;
  }
}

/**
 * Time-based circuit breaker mirroring the Rust `comms/backoff.rs` state
 * machine (Closed / Open / HalfOpen). A tripped breaker stays Open for
 * `cooldownMs`, then `allow(now)` promotes it to HalfOpen and lets a single
 * probe through — so an open breaker can heal itself instead of starving its
 * own reset (the previous boolean-`open` version could never recover without
 * an app restart, since `allow()` blocked the very attempt that would call
 * `onSuccess()`).
 */
export class CircuitBreaker {
  private state: CircuitState = "closed";
  private consecutiveFailures = 0;
  private consecutiveSuccesses = 0;
  private openedAt: number | null = null;
  private readonly failureThreshold: number;
  private readonly cooldownMs: number;
  private readonly successThreshold: number;

  constructor(config: CircuitBreakerConfig) {
    this.failureThreshold = config.failureThreshold;
    this.cooldownMs = config.cooldownMs ?? DEFAULT_BREAKER_COOLDOWN_MS;
    this.successThreshold = config.successThreshold ?? 1;
  }

  get currentState(): CircuitState {
    return this.state;
  }

  /** Back-compat accessor: true when the breaker is fully tripped. */
  get open(): boolean {
    return this.state === "open";
  }

  /**
   * Whether a request/attempt is permitted at `now`, transitioning
   * Open -> HalfOpen once the cooldown has elapsed.
   */
  allow(now: number = Date.now()): boolean {
    if (this.state === "open") {
      const elapsed = this.openedAt == null ? this.cooldownMs : now - this.openedAt;
      if (elapsed >= this.cooldownMs) {
        this.state = "half_open";
        this.consecutiveSuccesses = 0;
        return true;
      }
      return false;
    }
    return true;
  }

  onSuccess(): void {
    this.consecutiveFailures = 0;
    if (this.state === "half_open") {
      this.consecutiveSuccesses += 1;
      if (this.consecutiveSuccesses >= this.successThreshold) {
        this.state = "closed";
        this.consecutiveSuccesses = 0;
        this.openedAt = null;
      }
    } else if (this.state === "open") {
      this.state = "closed";
      this.openedAt = null;
    }
  }

  onFailure(now: number = Date.now()): void {
    this.consecutiveSuccesses = 0;
    if (this.state === "half_open") {
      this.trip(now);
    } else if (this.state === "closed") {
      this.consecutiveFailures += 1;
      if (this.consecutiveFailures >= this.failureThreshold) {
        this.trip(now);
      }
    }
  }

  /** Milliseconds until an Open breaker will permit a half-open probe. */
  remainingCooldownMs(now: number = Date.now()): number {
    if (this.state !== "open" || this.openedAt == null) return 0;
    return Math.max(0, this.cooldownMs - (now - this.openedAt));
  }

  /** Force back to a clean Closed state (deliberate reconnect / resume). */
  reset(): void {
    this.state = "closed";
    this.consecutiveFailures = 0;
    this.consecutiveSuccesses = 0;
    this.openedAt = null;
  }

  private trip(now: number): void {
    this.state = "open";
    this.openedAt = now;
    this.consecutiveFailures = 0;
    this.consecutiveSuccesses = 0;
  }
}

export type ReconnectSchedulerOptions = {
  policy: BackoffPolicy;
  failureThreshold?: number;
  onExhausted?: () => void;
};

/**
 * Schedules bounded reconnect attempts with overlap protection.
 * Used by workshop workspace + interactive SSE recovery loops.
 */
export class ReconnectScheduler {
  private attempt = 0;
  private timer: ReturnType<typeof setTimeout> | null = null;
  private readonly overlap = new OverlapGuard();
  private readonly breaker: CircuitBreaker;
  private readonly policy: BackoffPolicy;
  private readonly onExhausted?: () => void;
  private tornDown = false;

  constructor(options: ReconnectSchedulerOptions) {
    this.policy = options.policy;
    this.breaker = new CircuitBreaker({
      failureThreshold: options.failureThreshold ?? 5,
    });
    this.onExhausted = options.onExhausted;
  }

  get pending(): boolean {
    return this.timer != null;
  }

  cancel(): void {
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    this.attempt = 0;
    this.overlap.release();
    // A deliberate cancel (foreground resume, explicit reconnect, teardown) is a
    // clean slate — clear any tripped breaker so recovery isn't locked out.
    this.breaker.reset();
  }

  teardown(): void {
    this.tornDown = true;
    this.cancel();
  }

  /** Call after a successful reconnect to reset backoff. */
  noteSuccess(): void {
    this.attempt = 0;
    this.breaker.onSuccess();
  }

  /** Schedule `task` unless a timer is already pending or overlap rejects. */
  schedule(task: () => void | Promise<void>): void {
    if (this.tornDown || this.timer) return;

    const now = Date.now();
    const wasOpen = this.breaker.currentState === "open";
    if (!this.breaker.allow(now)) {
      // Breaker is Open and cooldown hasn't elapsed. Rather than dropping the
      // reconnect loop entirely (the old behavior — which meant a tripped
      // breaker never recovered without an app restart), arm a wake-up timer
      // for the remaining cooldown so a half-open probe fires automatically.
      this.armBreakerWakeup(task, now);
      return;
    }
    if (wasOpen) {
      // Just transitioned Open -> HalfOpen: this probe is a fresh chance, so
      // restart backoff (and re-arm the interactive attempt budget).
      this.attempt = 0;
    }
    if (!mayRetry(this.policy, this.attempt)) {
      this.onExhausted?.();
      return;
    }
    if (!this.overlap.tryEnter()) return;

    const delayMs = reconnectDelayMs(this.policy, this.attempt);
    this.attempt += 1;

    this.timer = setTimeout(() => {
      this.timer = null;
      void Promise.resolve(task())
        .catch(() => {
          this.breaker.onFailure();
        })
        .finally(() => {
          this.overlap.release();
        });
    }, delayMs);
  }

  private armBreakerWakeup(task: () => void | Promise<void>, now: number): void {
    const wait = Math.max(250, this.breaker.remainingCooldownMs(now));
    this.timer = setTimeout(() => {
      this.timer = null;
      this.schedule(task);
    }, wait);
  }
}

File: crates/medousa-sdk/src/reconnect.rs
//! Client-side reconnect discipline for interactive SSE streams.
//!
//! Mirrors the server `medousa::comms` backoff / circuit-breaker / overlap-guard
//! semantics so SDK consumers get bounded, jittered reattach with `?since=<seq>`
//! against the daemon's disk-backed turn spine.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use crate::transport::path_with_query;

/// Policy for reconnecting an interactive turn SSE stream after a drop.
#[derive(Debug, Clone)]
#[derive(Default)]
pub struct ReconnectPolicy {
    pub backoff: BackoffPolicy,
    pub breaker: CircuitBreakerConfig,
}


/// Exponential backoff with cap and deterministic pseudo-jitter (no `rand` dep).
#[derive(Debug, Clone)]
pub struct BackoffPolicy {
    pub base: Duration,
    pub factor: f64,
    pub max: Duration,
    pub max_attempts: Option<u32>,
}

impl Default for BackoffPolicy {
    fn default() -> Self {
        Self {
            base: Duration::from_millis(500),
            factor: 2.0,
            max: Duration::from_secs(30),
            max_attempts: Some(10),
        }
    }
}

impl BackoffPolicy {
    pub fn delay(&self, attempt: u32) -> Duration {
        let cap_ms = self.max.as_millis() as f64;
        let base_ms = self.base.as_millis() as f64;
        let factor = if self.factor < 1.0 { 1.0 } else { self.factor };
        let raw = base_ms * factor.powi(attempt.min(64) as i32);
        let clamped = raw.min(cap_ms).max(0.0);
        // Pseudo-jitter in `[0.5, 1.0]` of computed delay.
        let jitter = 0.5 + ((attempt as u64).wrapping_mul(7919) % 500) as f64 / 1000.0;
        Duration::from_millis((clamped * jitter) as u64)
    }

    pub fn may_retry(&self, attempt: u32) -> bool {
        match self.max_attempts {
            Some(max) => attempt < max,
            None => true,
        }
    }
}

#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    pub failure_threshold: u32,
    pub cooldown: Duration,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_threshold: 5,
            cooldown: Duration::from_secs(15),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitState {
    Closed,
    Open,
}

#[derive(Debug)]
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    consecutive_failures: u32,
    state: CircuitState,
}

impl CircuitBreaker {
    pub fn new(config: CircuitBreakerConfig) -> Self {
        Self {
            config,
            consecutive_failures: 0,
            state: CircuitState::Closed,
        }
    }

    pub fn state(&self) -> CircuitState {
        self.state
    }

    pub fn allow(&mut self) -> bool {
        self.state == CircuitState::Closed
    }

    pub fn on_success(&mut self) {
        self.consecutive_failures = 0;
        self.state = CircuitState::Closed;
    }

    pub fn on_failure(&mut self) {
        self.consecutive_failures = self.consecutive_failures.saturating_add(1);
        if self.consecutive_failures >= self.config.failure_threshold {
            self.state = CircuitState::Open;
        }
    }
}

/// Rejects a second concurrent reconnect on the same guard.
#[derive(Debug, Clone)]
pub struct OverlapGuard {
    active: Arc<AtomicBool>,
}

impl Default for OverlapGuard {
    fn default() -> Self {
        Self::new()
    }
}

impl OverlapGuard {
    pub fn new() -> Self {
        Self {
            active: Arc::new(AtomicBool::new(false)),
        }
    }

    pub fn try_enter(&self) -> Option<OverlapPermit> {
        if self
            .active
            .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
            .is_ok()
        {
            Some(OverlapPermit {
                active: Arc::clone(&self.active),
            })
        } else {
            None
        }
    }

    pub fn is_active(&self) -> bool {
        self.active.load(Ordering::Acquire)
    }
}

pub struct OverlapPermit {
    active: Arc<AtomicBool>,
}

impl Drop for OverlapPermit {
    fn drop(&mut self) {
        self.active.store(false, Ordering::Release);
    }
}

/// Strip any existing query string and append `?since=<seq>` for spine replay.
pub fn stream_path_with_since(path: &str, since: u64) -> String {
    let base = path.split('?').next().unwrap_or(path);
    if since == 0 {
        return path_with_query(base, &[]);
    }
    path_with_query(base, &[("since", since.to_string())])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stream_path_appends_since() {
        assert_eq!(
            stream_path_with_since("/v1/interactive/turn/t1/stream", 42),
            "/v1/interactive/turn/t1/stream?since=42"
        );
    }

    #[test]
    fn stream_path_replaces_existing_query() {
        assert_eq!(
            stream_path_with_since("/v1/interactive/turn/t1/stream?since=1", 99),
            "/v1/interactive/turn/t1/stream?since=99"
        );
    }

    #[test]
    fn backoff_caps_and_grows() {
        let policy = BackoffPolicy::default();
        assert!(policy.delay(0) >= Duration::from_millis(250));
        assert!(policy.delay(5) <= policy.max);
    }

    #[test]
    fn overlap_guard_admits_one() {
        let guard = OverlapGuard::new();
        let _a = guard.try_enter().expect("first");
        assert!(guard.try_enter().is_none());
    }
}
